src/Processors/QueryPlan/Optimizations/filterPushDown.cpp

```diff
@@ -204,6 +209,204 @@ static size_t simplePushDownOverStep(QueryPlan::Node * parent_node, QueryPlan::N
     return 0;
 }
 
+static size_t tryPushDownOverJoinStep(QueryPlan::Node * parent_node, QueryPlan::Nodes & nodes, QueryPlanStepPtr & child)
+{
+    auto & parent = parent_node->step;
+    auto * filter = assert_cast<FilterStep *>(parent.get());
+
+    auto * join = typeid_cast<JoinStep *>(child.get());
+    auto * filled_join = typeid_cast<FilledJoinStep *>(child.get());
+
+    if (!join && !filled_join)
+        return 0;
+
+    /** For equivalent JOIN with condition `ON lhs.x_1 = rhs.y_1 AND lhs.x_2 = rhs.y_2 ...`, we can build equivalent sets of columns and this
+      * will allow to push conditions that only use columns from equivalent sets to both sides of JOIN, without considering JOIN type.
+      *
+      * For example: `FROM lhs INNER JOIN rhs ON lhs.id = rhs.id AND lhs.value = rhs.value`
+      * In this example columns `id` and `value` from both tables are equivalent.
+      *
+      * During filter push down for different JOIN types filter push down logic is different:
+      *
+      * 1. For INNER JOIN we can push all valid conditions to both sides of JOIN. We also can push all valid conditions that use columns from
+      * equivalent sets to both sides of JOIN.
+      * 2. For LEFT/RIGHT JOIN we can push conditions that use columns from LEFT/RIGHT stream to LEFT/RIGHT JOIN side. We can also push conditions
+      * that use columns from LEFT/RIGHT equivalent sets to RIGHT/LEFT JOIN side.
+      *
+      * Additional filter push down optimizations:
+      * 1. TODO: Support building equivalent sets for more than 2 JOINS. It is possible, but will require more complex analysis step.
+      * 2. TODO: Support building equivalent sets for JOINs with more than 1 clause.
+      * 3. TODO: For LEFT/RIGHT JOIN, we can assume that RIGHT/LEFT columns used in filter will be default/NULL constants and
+      * check if filter will always be false, in those scenario we can transform LEFT/RIGHT JOIN into INNER JOIN and push conditions to both tables.
+      * 4. TODO: It is possible to pull up filter conditions from LEFT/RIGHT stream and push conditions that use columns from LEFT/RIGHT equivalent sets
+      * to RIGHT/LEFT JOIN side.
+      */
+
+    const auto & join_header = child->getOutputStream().header;
+    const auto & table_join = join ? join->getJoin()->getTableJoin() : filled_join->getJoin()->getTableJoin();
+    const auto & left_stream_input_header = child->getInputStreams().front().header;
+    const auto & right_stream_input_header = child->getInputStreams().back().header;
```

Comment:
In case of `FilledJoinStep`,  `sizeof(child->getInputStreams()) == 1` which `menas left_stream_input_header == right_stream_input_header`, it's an issue, althought I can't simply reproduce the issue, since `chooseJoinAlgorithm() in PlannerJoins.cpp` always call `TableJoin::setRename` to create aliases for columns in the right table

Reply:
Yes, it is possible to create crash with FilledJoinStep.
Fix https://github.com/ClickHouse/ClickHouse/pull/62624.